src/hotspot/cpu/x86/c1_LIRAssembler_x86.cpp

```diff
@@ -3458,6 +3488,11 @@ void LIR_Assembler::emit_arraycopy(LIR_OpArrayCopy* op) {
   bool disjoint = (flags & LIR_OpArrayCopy::overlapping) == 0;
   bool aligned = (flags & LIR_OpArrayCopy::unaligned) == 0;
   const char *name;
+  if (BarrierSet::barrier_set()->barrier_set_assembler()->enable_oop_arraycopy_prologue()
+    && (basic_type == T_OBJECT || basic_type == T_ARRAY)) {
+      // TODO: Support 32-bit x86
+      LP64_ONLY(if (c_rarg3 != dst) __ mov(c_rarg3, dst););
```

Comment:
First I cannot test the correctness of the 32-bit support because mmtk-openjdk only supports x64 now.

Actually this line has no issue on 32bit platforms. The problem is the checkcast-arraycopy function. It uses `c_rarg5`/`r9` to carry `dst_obj`. For 32-bit, this has to be spilled to the stack. After I submitted the PR for binding support, I will create an issue regarding this. But we can also choose to disable this API and use field barrier for array-copy.